plugins/FakeTpCreatorHeartbeatMaker.cpp

```diff
@@ -0,0 +1,178 @@
+      // The condition to exit the loop is that we've been stopped and
+      // there's nothing left on the input queue
+      if (!running_flag.load()) {
+        break;
+      } else {
+        continue;
+      }
+    }
+
+    TLOG_DEBUG(2) << "Activity received.";
+
+    dataformats::timestamp_t current_tpset_start_time = tpset.start_time;
+
+    bool send_heartbeat =
+      should_send_heartbeat(last_sent_heartbeat_time, current_tpset_start_time, is_first_tpset_received);
+
+    bool successfully_sent_real_tpset = false;
+    bool successfully_sent_heartbeat = false;
+
+    if (send_heartbeat) {
+      TPSet tpset_heartbeat;
+      get_heartbeat(tpset_heartbeat, current_tpset_start_time);
+      while (!successfully_sent_heartbeat) {
+        try {
+          m_output_queue->push(tpset_heartbeat, m_queue_timeout);
+          successfully_sent_heartbeat = true;
+          m_heartbeats_sent++;
+          last_sent_heartbeat_time = current_tpset_start_time;
+          is_first_tpset_received = false;
+        } catch (const dunedaq::appfwk::QueueTimeoutExpired& excpt) {
+          std::ostringstream oss_warn;
+          oss_warn << "push to output queue \"" << m_output_queue->get_name() << "\"";
+          ers::warning(
+            dunedaq::appfwk::QueueTimeoutExpired(ERS_HERE, get_name(), oss_warn.str(), m_queue_timeout.count()));
+        }
+      }
+    }
+    while (!successfully_sent_real_tpset) {
+      try {
+        m_output_queue->push(tpset, m_queue_timeout);
+        successfully_sent_real_tpset = true;
+        m_tpset_sent_count++;
+      } catch (const dunedaq::appfwk::QueueTimeoutExpired& excpt) {
+        std::ostringstream oss_warn;
+        oss_warn << "push to output queue \"" << m_output_queue->get_name() << "\"";
+        ers::warning(
+          dunedaq::appfwk::QueueTimeoutExpired(ERS_HERE, get_name(), oss_warn.str(), m_queue_timeout.count()));
+      }
+    }
+  }
+
+  TLOG() << "Received " << m_tpset_received_count << " and sent " << m_tpset_sent_count << " real TPSets. Sent "
+         << m_heartbeats_sent << " fake heartbeats." << std::endl;
+  TLOG_DEBUG(2) << "Exiting do_work() method";
+}
+
+bool
+FakeTpCreatorHeartbeatMaker::should_send_heartbeat(dataformats::timestamp_t const& last_sent_heartbeat_time,
+                                                   dataformats::timestamp_t const& current_tpset_start_time,
+                                                   bool const& is_first_tpset_received)
+{
+  // If it is the first TPSet received, send out a heartbeat.
+  // Else, can assume that the TPSets are already ordered by start_time. Therefore, check only
+  // that the difference between the start_time of the current TPSet and the time that the
+  // last heartbeat was sent out is greater than the specified heartbeat interval.
+  if (is_first_tpset_received)
+    return true;
+  else
+    return last_sent_heartbeat_time + m_heartbeat_interval > current_tpset_start_time;
```

Comment:
Is the ">" the right way round here? Say `last_sent_heartbeat_time` is 1000 and `m_heartbeat_interval` is 100, then we should send a heartbeat if  `current_tpset_start_time > 1100`, right?

Reply:
Yes, you're right. I'll fix that.